api/src/main/java/jakarta/servlet/http/Cookie.java

```diff
@@ -422,4 +423,103 @@ public void setHttpOnly(boolean isHttpOnly) {
+        this.sameSiteMode = sameSiteMode;
+    }
+
+    /**
+     * Returns the <i>SameSite</i>> mode associated with this cookie.
+     *
+     * @return The <i>SameSite</i> mode associated with this cookie. May be null.
+     *
+     * @since Servlet 4.x
+     */
+    public SameSiteMode getSameSiteMode() {
+        return sameSiteMode;
+    }
+
+    /**
+     * The available values that can be used with the cookie's <i>SameSite</i> attribute.
+     */
+    public enum SameSiteMode {
+
+        /**
+         * This mode effectively disables the protection provided by SameSite cookies.
+         *
+         * <p>
+         *     When SameSiteMode is {@literal None}, applications should have additional measures to protect cookies
+         *     from CSRF attacks.
+         * </p>
+         *
+         * <p>
+         *     Browsers may default to {@link SameSiteMode#LAX} or {@link SameSiteMode#STRICT} when the flag is not
+         *     explicitly set by applications in the Set-Cookie header. Setting the {@literal SameSite} flag to
+         *     {@literal None} ensures that this default behavior is disabled in browsers.
+         * </p>
+         *
+         * <p>
+         *     Note that this mode is browser dependent, and browsers may require cookies to have the {@literal Secure}
+         *     flag to be set at the same time.
+         * </p>
+         */
+        NONE("None"),
+        /**
+         * This mode allows sending cookies along for cross-site top level navigation requests.
+         *
+         * <p>
+         *     When SameSiteMode is {@literal Lax}, the browser is allowed to send the cookie value along with top level
+         *     navigation requests when making requests cross-site.
+         * </p>
+         */
+        LAX("Lax"),
+        /**
+         * This mode prevents browsers from sending cookies along with any kind of cross-site requests.
+         *
+         * <p>
+         *     When SameSiteMode is {@literal Strict}, the browser is not allowed to send the cookie value along with
+         *     cross-site requests.
+         * </p>
+         *
+         * <p>
+         *     This behaviour could prevent applications seeing the end-user's session when a cross-site navigation
+         *     takes place. To overcome this, applications should consider utilising multiple cookies to keep track of
+         *     sessions.
+         * </p>
+         */
+        STRICT("Strict");
+
+        private final String value;
+
+        SameSiteMode(String value) {
+            this.value = value;
+        }
+
+        /**
+         * Retrieves the {@link SameSiteMode} instance from its String representation.
+         *
+         * @param value The String to be converted.
+         * @return The corresponding {@link SameSiteMode} enum.
+         */
+        public static SameSiteMode fromString(String value) {
+            return valueOf(value.toUpperCase());
```

Comment:
To be on the safe side / for good practice (the fact that all values are case variations of enum value names is incidental), I suggest something like:
`return Stream.of(values()).filter(value::equals).findFirst().orElseThrow(IllegalArgumentException::new);`

Reply:
The code as written is mostly correct.
Only needs to change to using `value.toUpperCase(Locale.ENGLISH)` to avoid default locale behaviors.

Reply:
I'd like to stay pragmatic on this one and leave the code as is. If there is a requirement to add new values, the existing implementation can be revisited.

Reply:
@ymajoros the call to `valueOf(String)` in @aldaris's code already gives you the `IllegalArgumentException` and allows for case-insensitive lookup from spec recommended literal value too.

Example:

``` java
package enums;

import java.util.Locale;

public class SameSiteEnumDemo {
    public enum SameSiteMode {
        NONE("None"),
        LAX("Lax"),
        STRICT("Strict");

        private final String value;

        SameSiteMode(String value) { this.value = value; }

        public static SameSiteMode fromString(String value) {
            return valueOf(value.toUpperCase(Locale.ENGLISH));
        }
    }

    public static void main(String[] args) {
        testEnum("none");
        testEnum("LAX");
        testEnum("Bogus");
    }

    private static void testEnum(String str) {
        testFromString(str);
        testValueOf(str);
    }

    private static void testFromString(String str) {
        try {
            SameSiteMode ssm = SameSiteMode.fromString(str);
            System.out.printf("SameSiteMode.fromString(%s) = %s%n", str, ssm);
        } catch(Throwable cause) {
            System.err.printf("OOPS on fromString(%s) - %s: %s%n", str, cause.getClass().getName(), cause.getMessage());
        }
    }

    private static void testValueOf(String str) {
        try {
            SameSiteMode ssm = SameSiteMode.valueOf(str);
            System.out.printf("SameSiteMode.valueOf(%s) = %s%n", str, ssm);
        } catch(Throwable cause) {
            System.err.printf("OOPS on valueOf(%s) - %s: %s%n", str, cause.getClass().getName(), cause.getMessage());
        }
    }
}
```

With results of ...

```
SameSiteMode.fromString(none) = NONE
OOPS on valueOf(none) - java.lang.IllegalArgumentException: No enum constant enums.SameSiteEnumDemo.SameSiteMode.none
SameSiteMode.fromString(LAX) = LAX
SameSiteMode.valueOf(LAX) = LAX
OOPS on fromString(Bogus) - java.lang.IllegalArgumentException: No enum constant enums.SameSiteEnumDemo.SameSiteMode.BOGUS
OOPS on valueOf(Bogus) - java.lang.IllegalArgumentException: No enum constant enums.SameSiteEnumDemo.SameSiteMode.Bogus
```

Reply:
Exactly, in my understanding the problem suggested was that the enum name and the enum value does not always have to be the same and it would be more future-proof if the value would be checked exactly instead. My take is that there is no need to future-proof the code, because a) the spec will most likely not introduce new values, and b) the enums intentionally match with the enum values in the current implementation.

Reply:
If it has to be case insensitive (is it?), then you could still call toUpperCase() as well. I still think current's approach lowers the quality. Whatever, but then please close reviews so we don't waste time giving advice.